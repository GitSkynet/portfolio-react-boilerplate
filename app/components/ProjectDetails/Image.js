import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import ImageWrapper from './ImageWrapper';

const StyledPicture = styled.picture`
  * + * {
    margin: 0;
  }
`;

function transformImagesArray(images) {
  const srcSet = {};

  images.forEach((_image) => {
    const { file: { details: { image: { height, width } }, fileName, url } } = _image;
    const [relativeSizeIndication] = fileName
      .replace(/\.(png|jpeg)/i, '')
      .trim()
      .split('-')
      .slice(-1);
    srcSet[relativeSizeIndication] = {
      dimensions: {
        height,
        width,
      },
      url,
    };
  });

  return srcSet;
}

function getImageDescription(images) {
  const [{ title = '' } = {}] = images.slice(-1);
  return title;
}

function composeSourceElements(sources) {
  return sources.map((source) => {
    const { media: { breakpoint, size }, srcSet } = source;

    return (<source key={size} media={`(${breakpoint}: ${size})`} srcSet={srcSet} />);
  });
}

function Image({ images = [] }) {
  if (images.length === 0) {
    return <div />;
  }

  const { s, m, l, xl } = transformImagesArray(images);
  const altText = getImageDescription(images);
  const breakpoints = [
    {
      media: { breakpoint: 'max-width', size: '424px' },
      srcSet: s.url,
    },
    {
      media: { breakpoint: 'max-width', size: '767px' },
      srcSet: m.url,
    },
    {
      media: { breakpoint: 'max-width', size: '949px' },
      srcSet: l.url,
    },
    {
      media: { breakpoint: 'min-width', size: '950px' },
      srcSet: xl.url,
    },
  ];
  const imageDimensions = [s.dimensions, m.dimensions, l.dimensions, xl.dimensions];

  if (s.url && m.url && l.url && xl.url) {
    return (
      <ImageWrapper
        breakpoints={breakpoints}
        imageDimensions={imageDimensions}
      >
        <StyledPicture>
          {
            composeSourceElements(breakpoints)
          }
          <img src={xl.url} alt={altText} />
        </StyledPicture>
      </ImageWrapper>
    );
  }

  return <div />;
}

Image.propTypes = {
  images: PropTypes.arrayOf(
    PropTypes.shape({
      file: PropTypes.shape({
        details: PropTypes.shape({
          image: PropTypes.shape({
            height: PropTypes.number,
            width: PropTypes.number,
          }),
        }),
        fileName: PropTypes.string,
        url: PropTypes.string,
      }),
      title: PropTypes.string,
    })
  ),
};

export default Image;
