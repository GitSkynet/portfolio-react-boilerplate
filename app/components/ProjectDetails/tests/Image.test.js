import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Image from '../Image';

describe('<Image />', () => {
  const images = [
    {
      file: {
        details: {
          image: {
            height: 10,
            width: 20,
          },
        },
        fileName: 'test-s.png',
        url: '/test-s.png',
      },
      title: 'Test image',
    },
    {
      file: {
        details: {
          image: {
            height: 20,
            width: 30,
          },
        },
        fileName: 'test-m.png',
        url: '/test-m.png',
      },
      title: 'Test image',
    },
    {
      file: {
        details: {
          image: {
            height: 30,
            width: 40,
          },
        },
        fileName: 'test-l.png',
        url: '/test-l.png',
      },
      title: 'Test image',
    },
    {
      file: {
        details: {
          image: {
            height: 50,
            width: 60,
          },
        },
        fileName: 'test-xl.png',
        url: '/test-xl.png',
      },
      title: 'Test image',
    },
  ];

  it('should render and match the snapshot', () => {
    const renderedComponent = shallow(<Image images={images} />);
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });

  it('should render when no images are provided', () => {
    const renderedComponent = shallow(<Image images={[]} />);
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });
});
