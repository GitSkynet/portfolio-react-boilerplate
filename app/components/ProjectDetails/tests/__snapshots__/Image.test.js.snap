// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> should render and match the snapshot 1`] = `
<ImageWrapper__ImageWrapper
  breakpoints={
    Array [
      Object {
        "media": Object {
          "breakpoint": "max-width",
          "size": "424px",
        },
        "srcSet": "/test-s.png",
      },
      Object {
        "media": Object {
          "breakpoint": "max-width",
          "size": "767px",
        },
        "srcSet": "/test-m.png",
      },
      Object {
        "media": Object {
          "breakpoint": "max-width",
          "size": "949px",
        },
        "srcSet": "/test-l.png",
      },
      Object {
        "media": Object {
          "breakpoint": "min-width",
          "size": "950px",
        },
        "srcSet": "/test-xl.png",
      },
    ]
  }
  imageDimensions={
    Array [
      Object {
        "height": 10,
        "width": 20,
      },
      Object {
        "height": 20,
        "width": 30,
      },
      Object {
        "height": 30,
        "width": 40,
      },
      Object {
        "height": 50,
        "width": 60,
      },
    ]
  }
>
  <Image__StyledPicture>
    <source
      key="424px"
      media="(max-width: 424px)"
      srcSet="/test-s.png"
    />
    <source
      key="767px"
      media="(max-width: 767px)"
      srcSet="/test-m.png"
    />
    <source
      key="949px"
      media="(max-width: 949px)"
      srcSet="/test-l.png"
    />
    <source
      key="950px"
      media="(min-width: 950px)"
      srcSet="/test-xl.png"
    />
    <img
      alt="Test image"
      src="/test-xl.png"
    />
  </Image__StyledPicture>
</ImageWrapper__ImageWrapper>
`;

exports[`<Image /> should render when no images are provided 1`] = `<div />`;
